ConcurrentModificationException is thrown when the jsp files are compiled at the same time.

## Environment Details
* GlassFish Version (and build number): GlassFish 6.2.3
* JDK version: OpenJDK version 11.0.12
* OS: CentOS Linux release 7
* Database: none

----------

## Problem Description

500 error can occur when accessing a web application under the following conditions.  
1. jsp precompiling is unable.  
2. at the first time accessing.
3. when the jsp pages are accessed by multiple clients at the same time.

Then the following stack trace is printed in `server.log`. 
```
[2021-12-24T17:50:51.817+0900] [glassfish 6.2] [SEVERE] [] [] [tid: _ThreadID=29 _ThreadName=Thread-8] [timeMillis: 1640335851817] [levelValue: 1000] [[
  java.lang.RuntimeException: java.util.ConcurrentModificationException
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:163)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
        at org.apache.jasper.compiler.Jsr199JavaCompiler.compile(Jsr199JavaCompiler.java:213)
        at org.apache.jasper.compiler.Compiler.generateClass(Compiler.java:325)
        at org.apache.jasper.compiler.Compiler.compile(Compiler.java:387)
        at org.apache.jasper.JspCompilationContext.compile(JspCompilationContext.java:561)
        at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:288)
        at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:355)
        at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:284)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
        at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1540)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:217)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:119)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:611)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:550)
        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:75)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:114)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:332)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:199)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:440)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:144)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:174)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:153)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:196)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:88)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:246)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:178)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:118)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:96)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:51)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:510)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:82)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:83)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:34)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:101)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:535)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:515)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.util.ConcurrentModificationException
        at java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1493)
        at java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1521)
        at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper.wrapJavaFileObjects(ClientCodeWrapper.java:147)
        at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedJavaFileManager.list(ClientCodeWrapper.java:232)
        at jdk.compiler/com.sun.tools.javac.code.ClassFinder.list(ClassFinder.java:737)
        at jdk.compiler/com.sun.tools.javac.code.ClassFinder.scanUserPaths(ClassFinder.java:674)
        at jdk.compiler/com.sun.tools.javac.code.ClassFinder.fillIn(ClassFinder.java:551)
        at jdk.compiler/com.sun.tools.javac.code.ClassFinder.complete(ClassFinder.java:299)
        at jdk.compiler/com.sun.tools.javac.code.Symbol.complete(Symbol.java:642)
        at jdk.compiler/com.sun.tools.javac.code.Symbol$PackageSymbol.members(Symbol.java:1131)
        at jdk.compiler/com.sun.tools.javac.code.Symtab.listPackageModules(Symtab.java:832)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.visitTopLevel(Enter.java:345)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:529)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.classEnter(Enter.java:286)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.classEnter(Enter.java:301)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.complete(Enter.java:576)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.main(Enter.java:560)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.enterTrees(JavaCompiler.java:1066)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:937)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:147)
        ... 38 more
]]

[2021-12-24T17:50:51.850+0900] [glassfish 6.2] [WARNING] [] [jakarta.enterprise.web] [tid: _ThreadID=29 _ThreadName=http-listener-1(1)] [timeMillis: 1640335851850] [levelValue: 900] [[
  StandardWrapperValve[jsp]: Servlet.service() for servlet jsp threw exception
java.util.ConcurrentModificationException
        at java.base/java.util.HashMap$HashIterator.nextNode(HashMap.java:1493)
        at java.base/java.util.HashMap$ValueIterator.next(HashMap.java:1521)
        at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper.wrapJavaFileObjects(ClientCodeWrapper.java:147)
        at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedJavaFileManager.list(ClientCodeWrapper.java:232)
        at jdk.compiler/com.sun.tools.javac.code.ClassFinder.list(ClassFinder.java:737)
        at jdk.compiler/com.sun.tools.javac.code.ClassFinder.scanUserPaths(ClassFinder.java:674)
        at jdk.compiler/com.sun.tools.javac.code.ClassFinder.fillIn(ClassFinder.java:551)
        at jdk.compiler/com.sun.tools.javac.code.ClassFinder.complete(ClassFinder.java:299)
        at jdk.compiler/com.sun.tools.javac.code.Symbol.complete(Symbol.java:642)
        at jdk.compiler/com.sun.tools.javac.code.Symbol$PackageSymbol.members(Symbol.java:1131)
        at jdk.compiler/com.sun.tools.javac.code.Symtab.listPackageModules(Symtab.java:832)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.visitTopLevel(Enter.java:345)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept(JCTree.java:529)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.classEnter(Enter.java:286)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.classEnter(Enter.java:301)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.complete(Enter.java:576)
        at jdk.compiler/com.sun.tools.javac.comp.Enter.main(Enter.java:560)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.enterTrees(JavaCompiler.java:1066)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:937)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:147)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
        at org.apache.jasper.compiler.Jsr199JavaCompiler.compile(Jsr199JavaCompiler.java:213)
        at org.apache.jasper.compiler.Compiler.generateClass(Compiler.java:325)
        at org.apache.jasper.compiler.Compiler.compile(Compiler.java:387)
        at org.apache.jasper.JspCompilationContext.compile(JspCompilationContext.java:561)
        at org.apache.jasper.servlet.JspServletWrapper.service(JspServletWrapper.java:288)
        at org.apache.jasper.servlet.JspServlet.serviceJspFile(JspServlet.java:355)
        at org.apache.jasper.servlet.JspServlet.service(JspServlet.java:284)
        at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
        at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1540)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:217)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:119)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:611)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:550)
        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:75)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:114)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:332)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:199)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:440)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:144)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:174)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:153)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:196)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:88)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:246)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:178)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:118)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:96)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:51)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:510)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:82)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:83)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:34)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:101)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:535)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:515)
        at java.base/java.lang.Thread.run(Thread.java:829)
]]
```
The exception is occurred in following line:  
https://github.com/eclipse-ee4j/wasp/blob/fabca4d0ec26d8cf5ba5ad17302fddf7630d9811/src/main/java/org/glassfish/wasp/compiler/Jsr199JavaCompiler.java#L312  
The `packageFiles` can be accessed from multiple threads, so should be ConcurrentHashMap.  

## Steps to reproduce
To reproduce this problem more reliably, we need to use a debugger as follows:
1. prepare jsp app.  
[test.zip](https://github.com/eclipse-ee4j/wasp/files/7773584/test.zip)
2. execute commands,   
`asadmin start-domain --debug`  
`asadmin deploy test.war`  
3. access `{IP address}:{port}/test/test.jsp`, open a page which is shown `This is a test.jsp page.`.
4. use Eclipse IDE. start remote debug.  
5. set a break point at 
https://github.com/eclipse-ee4j/wasp/blob/fabca4d0ec26d8cf5ba5ad17302fddf7630d9811/src/main/java/org/glassfish/wasp/compiler/Jsr199JavaCompiler.java#L312
6. access `{IP address}:{port}/test/test2.jsp`, then stop at the break point.  
7. remove the break point.
8. proceed the step until arriving `Map.next()`.  
9. access `{IP address}:{port}/test/test3.jsp`, open a page which is shown `This is a test3.jsp page.`.
10. resume debugger.  
11. the page which accessed `{IP address}:{port}/test/test2.jsp` is shown 500 error.

--- a/src/main/java/org/glassfish/wasp/compiler/Jsr199JavaCompiler.java
+++ b/src/main/java/org/glassfish/wasp/compiler/Jsr199JavaCompiler.java
@@ -1,5 +1,6 @@
 /*
  * Copyright (c) 1997, 2020 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Public License v. 2.0, which is available at
@@ -34,10 +35,10 @@ import java.io.Writer;
 import java.net.URI;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 
 import javax.tools.Diagnostic;
 import javax.tools.DiagnosticCollector;
@@ -273,7 +274,7 @@ public class Jsr199JavaCompiler implements JavaCompiler {
         Map<String, Map<String, JavaFileObject>> packageMap = rtctxt.getPackageMap();
         Map<String, JavaFileObject> packageFiles = packageMap.get(packageName);
         if (packageFiles == null) {
-            packageFiles = new HashMap<>();
+            packageFiles = new ConcurrentHashMap<>();
             packageMap.put(packageName, packageFiles);
         }
         packageFiles.put(className, classFile);

